File: Power Query/300-399/384/PQ_384.py
import pandas as pd
import numpy as np

path = "300-399/384/PQ_Challenge_384.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=20)
test = pd.read_excel(path, usecols="F:H", nrows=66)

input["Update Date"] = pd.to_datetime(input["Update Date"], dayfirst=True)

result = (
    pd.MultiIndex.from_product(
        [pd.date_range(input["Update Date"].min(), input["Update Date"].max(), freq="D"),
         input["Product"].unique()],
        names=["Date", "Product"]
    ).to_frame(index=False)
    .merge(input, left_on=["Date", "Product"], right_on=["Update Date", "Product"], how="left")
    .sort_values(["Product", "Date"])
    .assign(
        update_flag=lambda x: x["Confirmed Price"].notna(),
        Confirmed_Price=lambda x: x.groupby("Product")["Confirmed Price"].ffill(),
        grp=lambda x: x["Confirmed Price"].notna().groupby(x["Product"]).cumsum()
    )
)
result["day"] = result.groupby(["Product", "grp"]).cumcount() + 1
result["Adjusted Price"] = np.maximum(
    result.groupby(["Product", "grp"])["Confirmed_Price"].transform("first")
    * (1 - 0.1 * np.maximum(result["day"] - 5, 0)),
    0
)
result = result[["Product", "Date", "Adjusted Price"]].rename(columns = {"Product":"Product.1"}).reset_index(drop=True)

print(result.equals(test))
# True

File: Power Query/300-399/384/PQ_384.R
library(tidyverse)
library(readxl)
library(lubridate)

path <- "300-399/384/PQ_Challenge_384.xlsx"
input <- read_excel(path, range = "A1:C21")
test <- read_excel(path, range = "F1:H66")

calendar = expand.grid(
  Date = seq(
    as.Date(min(input$`Update Date`, na.rm = TRUE)),
    as.Date(max(input$`Update Date`, na.rm = TRUE)),
    by = "days"
  ),
  Product = unique(input$Product)
)

result = calendar %>%
  left_join(input, by = c("Date" = "Update Date", "Product")) %>%
  arrange(Product, Date) %>%
  group_by(Product) %>%
  mutate(update_flag = !is.na(`Confirmed Price`)) %>%
  fill(`Confirmed Price`, .direction = "down") %>%
  mutate(grp = cumsum(update_flag)) %>%
  group_by(Product, grp) %>%
  mutate(
    day = row_number(),
    base = first(`Confirmed Price`),
    decay = pmax(day - 5, 0),
    `Adjusted Price` = pmax(base * (1 - 0.1 * decay), 0)
  ) %>%
  ungroup() %>%
  select(Product, Date, `Adjusted Price`)

all.equal(result, test)
# [1] TRUE
